services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java

```diff
@@ -1141,24 +1141,25 @@ public AccessTokenResponseBuilder generateRefreshToken(String scope) {
         }
 
         private void generateRefreshToken(boolean offlineTokenRequested) {
+            refreshToken = new RefreshToken(accessToken);
+            refreshToken.id(KeycloakModelUtils.generateId());
+            refreshToken.issuedNow();
+            clientSessionCtx.getClientSession().setTimestamp(Time.currentTime());
```

Comment:
@pedroigor I don't know much about authz but in general refresh token the user and client session timestamps are updated in [refreshAccessToken](https://github.com/keycloak/keycloak/blob/37cdc95bf3cc741b64fc3666233ccc29793444fd/services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java#L431-L433). Probably in the authz execution path that method is not called, but then I think that updating `lastSessionRefresh` in the user session is also needed (check if it's needed or it's done in another place).

Reply:
We don't and the behavior is that you would need a refresh token so they don't become idle.

But regarding the change, doesn't make more sense to just ise the `iat` to calculate the expiration time? In this method we are creating a refresh token so why use the timestamp from the client session?

Reply:
Also, whenever we issue a refresh token, the `lastSessionRefresh` should also be updated in the user session, no? 

That is what I'm trying to point out here ... We are actually issuing a new set of tokens so the sessions should not get idle if you are operating in the same (or a new) user/client session.

Reply:
@pedroigor Yes, if a refresh token is issued both sessions timestamps are updated, but right now for normal refresh is done in the other method [refreshAccessToken](https://github.com/keycloak/keycloak/blob/37cdc95bf3cc741b64fc3666233ccc29793444fd/services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java#L431-L433). I suppose in authz the other method is not executed, right?

In that case I think you also need to update the user session lastSessionRefresh like in the other method.

Reply:
@rmartinc Just update the PR to make clear what I'm trying to point out. Based on what you are saying, we should update the sessions (both user and client) whenever a refresh token is issued/created.

I've also updated this method https://github.com/keycloak/keycloak/pull/27577/files#diff-cbee120571b12f77f69db4cb51c259afea6d79df968c65b9879ab0b633b3db4fR142 to always take the `iat` as a input for the expiration rather than the "lastRefreshed". AFAIK, that was how the expiration was calculated before (not using the session timestamp) where the current time was used. Does this part make sense for you too?